warehouse/query-core/src/main/java/datawave/query/data/parsers/FieldIndexKey.java

```diff
@@ -0,0 +1,116 @@
+package datawave.query.data.parsers;
+
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Key;
+
+/**
+ * A reusable, dedicated key parser for Field Index keys
+ * <p>
+ * A Field Index key's components are
+ * <ul>
+ * <li>row - the shard</li>
+ * <li>column family - "fi\0FIELD_NAME"</li>
+ * <li>column qualifier - "value\0datatype\0uid"</li>
+ * </ul>
+ */
+public class FieldIndexKey {
+    
+    private String field;
+    private String value;
+    private String datatype;
+    private String uid;
+    
+    private ByteSequence cqBytes;
+    
+    private int firstNull;
+    private int secondNull;
+    
+    private Key key;
+    
+    /**
+     * Sets the key and resets all supporting objects
+     *
+     * @param k
+     *            a field index key
+     */
+    public void parse(Key k) {
+        this.key = k;
+        clearState();
```

Comment:
Should we be validating this is an fi key?

Reply:
The FieldIndexKey contract assumes it is operating on FI keys, that is, it is up to the caller to verify all keys passing through the parser are formatted properly. Where the parser is integrated into the IteratorBuildingVisitor (and by extension the ivarators) all of those ranges are bounded to the FI. 

Minimal effort is made to validate proper key structure. If an invalid key structure is detected, a null value is returned. If the caller blindly trusts the parser, this will turn into a null pointer exception.